File: plugins/plugin-browser/typescript/src/services/websocket-client.ts
import { logger } from "@elizaos/core";
import WebSocket from "ws";
import type { NavigationResult, WebSocketMessage, WebSocketResponse } from "../types.js";

type MessageHandler = (response: WebSocketResponse) => void;

export class BrowserWebSocketClient {
  private ws: WebSocket | null = null;
  private messageHandlers = new Map<string, MessageHandler>();
  private connected = false;
  private reconnectAttempts = 0;
  private maxReconnectAttempts = 5;
  private reconnectDelay = 1000;

  constructor(private serverUrl: string) {}

  async connect(): Promise<void> {
    return new Promise((resolve, reject) => {
      try {
        this.ws = new WebSocket(this.serverUrl);

        this.ws.on("open", () => {
          this.connected = true;
          this.reconnectAttempts = 0;
          logger.info(`[Browser] Connected to server at ${this.serverUrl}`);
          resolve();
        });

        this.ws.on("message", (data: WebSocket.RawData) => {
          try {
            const messageText = this.rawDataToString(data);
            const message = JSON.parse(messageText) as WebSocketResponse;

            if (message.requestId && this.messageHandlers.has(message.requestId)) {
              const handler = this.messageHandlers.get(message.requestId);
              this.messageHandlers.delete(message.requestId);
              handler?.(message);
            }

            if (message.type === "connected") {
              logger.info(`[Browser] Server connected: ${JSON.stringify(message)}`);
            }
          } catch (error) {
            const errorMessage = error instanceof Error ? error.message : String(error);
            logger.error(`[Browser] Error parsing message: ${errorMessage}`);
          }
        });

        this.ws.on("error", (error: Error) => {
          logger.error(`[Browser] WebSocket error: ${error.message}`);
          if (!this.connected) {
            reject(error);
          }
        });

        this.ws.on("close", () => {
          this.connected = false;
          logger.info("[Browser] Disconnected from server");

          if (this.ws && this.reconnectAttempts < this.maxReconnectAttempts) {
            this.attemptReconnect();
          }
        });
      } catch (error) {
        reject(error);
      }
    });
  }

  private rawDataToString(data: WebSocket.RawData): string {
    if (typeof data === "string") {
      return data;
    }
    if (Buffer.isBuffer(data)) {
      return data.toString("utf8");
    }
    if (data instanceof ArrayBuffer) {
      return Buffer.from(data).toString("utf8");
    }
    if (Array.isArray(data)) {
      return Buffer.concat(data).toString("utf8");
    }
    return String(data);
  }

  private async attemptReconnect(): Promise<void> {
    this.reconnectAttempts++;
    logger.info(
      `[Browser] Attempting reconnection ${this.reconnectAttempts}/${this.maxReconnectAttempts}...`
    );

    await new Promise((resolve) =>
      setTimeout(resolve, this.reconnectDelay * this.reconnectAttempts)
    );

    try {
      await this.connect();
    } catch (error) {
      const errorMessage = error instanceof Error ? error.message : String(error);
      logger.error(`[Browser] Reconnection failed: ${errorMessage}`);
    }
  }

  async sendMessage(type: string, data: Record<string, unknown>): Promise<WebSocketResponse> {
    if (!this.ws || !this.connected) {
      throw new Error("Not connected to browser server");
    }

    const requestId = `req-${Date.now()}-${Math.random().toString(36).substring(7)}`;
    const message: WebSocketMessage = {
      type,
      requestId,
      ...data,
    };

    return new Promise((resolve, reject) => {
      const timeout = setTimeout(() => {
        this.messageHandlers.delete(requestId);
        reject(new Error(`Request timeout for ${type}`));
      }, 30000);

      this.messageHandlers.set(requestId, (response) => {
        clearTimeout(timeout);
        if (response.type === "error") {
          reject(new Error(response.error ?? "Unknown error"));
        } else {
          resolve(response);
        }
      });

      this.ws?.send(JSON.stringify(message));
      logger.debug(`[Browser] Sent message: ${type} (${requestId})`);
    });
  }

  disconnect(): void {
    this.reconnectAttempts = this.maxReconnectAttempts;
    if (this.ws) {
      this.ws.close();
      this.ws = null;
    }
    this.connected = false;
    logger.info("[Browser] Client disconnected");
  }

  isConnected(): boolean {
    return this.connected;
  }

  async navigate(sessionId: string, url: string): Promise<NavigationResult> {
    const response = await this.sendMessage("navigate", {
      sessionId,
      data: { url },
    });
    const data = response.data as Record<string, unknown> | undefined;
    return {
      success: Boolean(data?.success),
      url: String(data?.url ?? url),
      title: String(data?.title ?? ""),
    };
  }

  async getState(
    sessionId: string
  ): Promise<{ url: string; title: string; sessionId: string; createdAt: Date }> {
    const response = await this.sendMessage("getState", { sessionId });
    const data = response.data as Record<string, unknown> | undefined;
    return {
      url: String(data?.url ?? ""),
      title: String(data?.title ?? ""),
      sessionId,
      createdAt: new Date(),
    };
  }

  async goBack(sessionId: string): Promise<NavigationResult> {
    const response = await this.sendMessage("goBack", { sessionId });
    const data = response.data as Record<string, unknown> | undefined;
    return {
      success: Boolean(data?.success ?? true),
      url: String(data?.url ?? ""),
      title: String(data?.title ?? ""),
    };
  }

  async goForward(sessionId: string): Promise<NavigationResult> {
    const response = await this.sendMessage("goForward", { sessionId });
    const data = response.data as Record<string, unknown> | undefined;
    return {
      success: Boolean(data?.success ?? true),
      url: String(data?.url ?? ""),
      title: String(data?.title ?? ""),
    };
  }

  async refresh(sessionId: string): Promise<NavigationResult> {
    const response = await this.sendMessage("refresh", { sessionId });
    const data = response.data as Record<string, unknown> | undefined;
    return {
      success: Boolean(data?.success ?? true),
      url: String(data?.url ?? ""),
      title: String(data?.title ?? ""),
    };
  }

  async click(sessionId: string, description: string): Promise<WebSocketResponse> {
    return await this.sendMessage("click", {
      sessionId,
      data: { description },
    });
  }

  async type(sessionId: string, text: string, field: string): Promise<WebSocketResponse> {
    return await this.sendMessage("type", {
      sessionId,
      data: { text, field },
    });
  }

  async select(sessionId: string, option: string, dropdown: string): Promise<WebSocketResponse> {
    return await this.sendMessage("select", {
      sessionId,
      data: { option, dropdown },
    });
  }

  async extract(sessionId: string, instruction: string): Promise<WebSocketResponse> {
    return await this.sendMessage("extract", {
      sessionId,
      data: { instruction },
    });
  }

  async screenshot(sessionId: string): Promise<WebSocketResponse> {
    return await this.sendMessage("screenshot", { sessionId });
  }

  async solveCaptcha(sessionId: string): Promise<WebSocketResponse> {
    return await this.sendMessage("solveCaptcha", { sessionId });
  }

  async health(): Promise<boolean> {
    try {
      const response = await this.sendMessage("health", {});
      return response.type === "health" && (response.data as { status: string })?.status === "ok";
    } catch (error) {
      const errorMessage = error instanceof Error ? error.message : String(error);
      logger.error(`[Browser] Health check failed: ${errorMessage}`);
      return false;
    }
  }
}

File: plugins/plugin-browser/rust/src/services/websocket_client.rs
use crate::types::{NavigationResult, WebSocketMessage, WebSocketResponse};
use futures_util::{SinkExt, StreamExt};
use std::collections::HashMap;
use std::sync::Arc;
use tokio::sync::{mpsc, RwLock};
use tokio_tungstenite::{connect_async, tungstenite::Message};
use tracing::{debug, error, info};
use uuid::Uuid;

type ResponseSender = mpsc::Sender<WebSocketResponse>;
type ResponseMap = Arc<RwLock<HashMap<String, ResponseSender>>>;

pub struct BrowserWebSocketClient {
    server_url: String,
    connected: Arc<RwLock<bool>>,
    pending_requests: ResponseMap,
    sender: Arc<RwLock<Option<mpsc::Sender<Message>>>>,
}

impl BrowserWebSocketClient {
    pub fn new(server_url: &str) -> Self {
        Self {
            server_url: server_url.to_string(),
            connected: Arc::new(RwLock::new(false)),
            pending_requests: Arc::new(RwLock::new(HashMap::new())),
            sender: Arc::new(RwLock::new(None)),
        }
    }

    pub async fn connect(&self) -> Result<(), String> {
        let url = url::Url::parse(&self.server_url).map_err(|e| format!("Invalid URL: {}", e))?;

        let (ws_stream, _) = connect_async(url)
            .await
            .map_err(|e| format!("Connection failed: {}", e))?;

        let (mut write, mut read) = ws_stream.split();

        let (tx, mut rx) = mpsc::channel::<Message>(100);
        *self.sender.write().await = Some(tx);
        *self.connected.write().await = true;

        info!("[Browser] Connected to server at {}", self.server_url);

        let pending = Arc::clone(&self.pending_requests);
        let connected = Arc::clone(&self.connected);

        tokio::spawn(async move {
            while let Some(msg) = read.next().await {
                match msg {
                    Ok(Message::Text(text)) => {
                        if let Ok(response) = serde_json::from_str::<WebSocketResponse>(&text) {
                            let requests = pending.read().await;
                            if let Some(sender) = requests.get(&response.request_id) {
                                let _ = sender.send(response).await;
                            }
                        }
                    }
                    Ok(Message::Close(_)) => {
                        info!("[Browser] Connection closed");
                        *connected.write().await = false;
                        break;
                    }
                    Err(e) => {
                        error!("[Browser] Error receiving message: {}", e);
                    }
                    _ => {}
                }
            }
        });

        tokio::spawn(async move {
            while let Some(msg) = rx.recv().await {
                if let Err(e) = write.send(msg).await {
                    error!("[Browser] Error sending message: {}", e);
                    break;
                }
            }
        });

        Ok(())
    }

    pub async fn send_message(
        &self,
        msg_type: &str,
        data: HashMap<String, serde_json::Value>,
    ) -> Result<WebSocketResponse, String> {
        if !*self.connected.read().await {
            return Err("Not connected to browser server".to_string());
        }

        let request_id = Uuid::new_v4().to_string();
        let mut message_data = data;

        let message = WebSocketMessage {
            msg_type: msg_type.to_string(),
            request_id: request_id.clone(),
            session_id: message_data
                .remove("sessionId")
                .and_then(|v| v.as_str().map(|s| s.to_string())),
            data: if message_data.is_empty() {
                None
            } else {
                Some(message_data)
            },
        };

        let json = serde_json::to_string(&message)
            .map_err(|e| format!("Failed to serialize message: {}", e))?;

        let (tx, mut rx) = mpsc::channel::<WebSocketResponse>(1);
        self.pending_requests
            .write()
            .await
            .insert(request_id.clone(), tx);

        if let Some(sender) = self.sender.read().await.as_ref() {
            sender
                .send(Message::Text(json))
                .await
                .map_err(|e| format!("Failed to send message: {}", e))?;
        }

        debug!("[Browser] Sent message: {} ({})", msg_type, request_id);

        let response = tokio::time::timeout(std::time::Duration::from_secs(30), rx.recv())
            .await
            .map_err(|_| format!("Request timeout for {}", msg_type))?
            .ok_or_else(|| "No response received".to_string())?;

        self.pending_requests.write().await.remove(&request_id);

        if response.msg_type == "error" {
            return Err(response
                .error
                .unwrap_or_else(|| "Unknown error".to_string()));
        }

        Ok(response)
    }

    pub async fn disconnect(&self) {
        *self.connected.write().await = false;
        *self.sender.write().await = None;
        info!("[Browser] Client disconnected");
    }

    pub async fn is_connected(&self) -> bool {
        *self.connected.read().await
    }

    pub async fn navigate(&self, session_id: &str, url: &str) -> Result<NavigationResult, String> {
        let mut data = HashMap::new();
        data.insert("sessionId".to_string(), serde_json::json!(session_id));

        let mut inner_data = HashMap::new();
        inner_data.insert("url".to_string(), serde_json::json!(url));
        data.insert("data".to_string(), serde_json::json!(inner_data));

        let response = self.send_message("navigate", data).await?;

        let resp_data = response.data.unwrap_or_default();
        Ok(NavigationResult {
            success: response.success,
            url: resp_data
                .get("url")
                .and_then(|v| v.as_str())
                .unwrap_or(url)
                .to_string(),
            title: resp_data
                .get("title")
                .and_then(|v| v.as_str())
                .unwrap_or("")
                .to_string(),
            error: response.error,
        })
    }

    pub async fn get_state(
        &self,
        session_id: &str,
    ) -> Result<HashMap<String, serde_json::Value>, String> {
        let mut data = HashMap::new();
        data.insert("sessionId".to_string(), serde_json::json!(session_id));

        let response = self.send_message("getState", data).await?;
        Ok(response.data.unwrap_or_default())
    }

    pub async fn go_back(&self, session_id: &str) -> Result<NavigationResult, String> {
        let mut data = HashMap::new();
        data.insert("sessionId".to_string(), serde_json::json!(session_id));

        let response = self.send_message("goBack", data).await?;
        let resp_data = response.data.unwrap_or_default();

        Ok(NavigationResult {
            success: response.success,
            url: resp_data
                .get("url")
                .and_then(|v| v.as_str())
                .unwrap_or("")
                .to_string(),
            title: resp_data
                .get("title")
                .and_then(|v| v.as_str())
                .unwrap_or("")
                .to_string(),
            error: response.error,
        })
    }

    pub async fn go_forward(&self, session_id: &str) -> Result<NavigationResult, String> {
        let mut data = HashMap::new();
        data.insert("sessionId".to_string(), serde_json::json!(session_id));

        let response = self.send_message("goForward", data).await?;
        let resp_data = response.data.unwrap_or_default();

        Ok(NavigationResult {
            success: response.success,
            url: resp_data
                .get("url")
                .and_then(|v| v.as_str())
                .unwrap_or("")
                .to_string(),
            title: resp_data
                .get("title")
                .and_then(|v| v.as_str())
                .unwrap_or("")
                .to_string(),
            error: response.error,
        })
    }

    pub async fn refresh(&self, session_id: &str) -> Result<NavigationResult, String> {
        let mut data = HashMap::new();
        data.insert("sessionId".to_string(), serde_json::json!(session_id));

        let response = self.send_message("refresh", data).await?;
        let resp_data = response.data.unwrap_or_default();

        Ok(NavigationResult {
            success: response.success,
            url: resp_data
                .get("url")
                .and_then(|v| v.as_str())
                .unwrap_or("")
                .to_string(),
            title: resp_data
                .get("title")
                .and_then(|v| v.as_str())
                .unwrap_or("")
                .to_string(),
            error: response.error,
        })
    }

    pub async fn click(
        &self,
        session_id: &str,
        description: &str,
    ) -> Result<WebSocketResponse, String> {
        let mut data = HashMap::new();
        data.insert("sessionId".to_string(), serde_json::json!(session_id));

        let mut inner_data = HashMap::new();
        inner_data.insert("description".to_string(), serde_json::json!(description));
        data.insert("data".to_string(), serde_json::json!(inner_data));

        self.send_message("click", data).await
    }

    pub async fn type_text(
        &self,
        session_id: &str,
        text: &str,
        field: &str,
    ) -> Result<WebSocketResponse, String> {
        let mut data = HashMap::new();
        data.insert("sessionId".to_string(), serde_json::json!(session_id));

        let mut inner_data = HashMap::new();
        inner_data.insert("text".to_string(), serde_json::json!(text));
        inner_data.insert("field".to_string(), serde_json::json!(field));
        data.insert("data".to_string(), serde_json::json!(inner_data));

        self.send_message("type", data).await
    }

    pub async fn select(
        &self,
        session_id: &str,
        option: &str,
        dropdown: &str,
    ) -> Result<WebSocketResponse, String> {
        let mut data = HashMap::new();
        data.insert("sessionId".to_string(), serde_json::json!(session_id));

        let mut inner_data = HashMap::new();
        inner_data.insert("option".to_string(), serde_json::json!(option));
        inner_data.insert("dropdown".to_string(), serde_json::json!(dropdown));
        data.insert("data".to_string(), serde_json::json!(inner_data));

        self.send_message("select", data).await
    }

    pub async fn extract(
        &self,
        session_id: &str,
        instruction: &str,
    ) -> Result<WebSocketResponse, String> {
        let mut data = HashMap::new();
        data.insert("sessionId".to_string(), serde_json::json!(session_id));

        let mut inner_data = HashMap::new();
        inner_data.insert("instruction".to_string(), serde_json::json!(instruction));
        data.insert("data".to_string(), serde_json::json!(inner_data));

        self.send_message("extract", data).await
    }

    pub async fn screenshot(&self, session_id: &str) -> Result<WebSocketResponse, String> {
        let mut data = HashMap::new();
        data.insert("sessionId".to_string(), serde_json::json!(session_id));

        self.send_message("screenshot", data).await
    }

    pub async fn solve_captcha(&self, session_id: &str) -> Result<WebSocketResponse, String> {
        let mut data = HashMap::new();
        data.insert("sessionId".to_string(), serde_json::json!(session_id));

        self.send_message("solveCaptcha", data).await
    }

    pub async fn health(&self) -> Result<bool, String> {
        let response = self.send_message("health", HashMap::new()).await?;

        let is_healthy = response.msg_type == "health"
            && response
                .data
                .as_ref()
                .and_then(|d| d.get("status"))
                .and_then(|v| v.as_str())
                == Some("ok");

        Ok(is_healthy)
    }
}
